### ISSUE TITLE
FR:  Periodically poll recent Discord messages and insert anecdotes unprompted

### ISSUE THREAD

**Issue by andrewmortiz** [jlyons210/discord-bot-ol-bootsie#13]
Possibly something along the lines of...based on ${RECENT_CONVERSATION} add your own anecdote to the conversation.  Could possibly have it rotate from anecdotes to questions, thoughts, ect, to diversify the additions.

**Comment by jlyons210** [comment]
Adding some additional notes. I am now capturing all message history for the duration of `BOT_THREAD_RETAIN_SEC`, and store in the `messageHistory` a `.isPromptContext` (bool) value (true = bot was @-mentioned, false = channel chatter).

---

The examples below open up some possibilities for unprompted bot interactions with a channel.

Asking the OpenAI API for single-word responses for mood/tone/sentiment of a statement unlocks a lot of possibilities.

Possibilities could include random interjections - which would require evaluating the collection of previous messages, and reactions - which would only need the most recent comment. I could also adjust the `system` prompt live if someone is acting extremely.

To prevent the bot from spamming, a new environment variable should be introduced, like `BOT_ENGAGE_PCT` to give "odds" of comments being engaged on.

Examples:
> @Ol' Bootsie describe the mood of this statement in one word: "Hey there, big boy. Want to go on a date?"

> Ol' Bootsie:
> Mood: Flirtatious
> Tone: Playful
> Sentiment: Positive

> @Ol' Bootsie do it again for the following statement: "Why did the tomato turn red? Because it saw the salad dressing! Meow!"

> Ol' Bootsie:
> Mood: Humorous
> Tone: Silly
> Sentiment: Positive

> @Ol' Bootsie do it again for the following statement: "I have the worst headache, I'm tired, and I want to go to bed."

> Ol' Bootsie:
> Mood: Unwell
> Tone: Fatigued
> Sentiment: Negative

**Comment by jlyons210** [comment]
PR #27 adds sentiment analysis to `HistoryMessage` objects (used in the `messageHistory[]` collection). Laying some groundwork.

**Comment by jlyons210** [comment]
User privacy may not be a huge concern, as of Mar 1, 2023. According to the [OpenAI Privacy Policy](https://openai.com/policies/api-data-usage-policies):
>Starting on March 1, 2023, we are making two changes to our data usage and retention policies:
>1. OpenAI will not use data submitted by customers via our API to train or improve our models, unless you explicitly decide to share your data with us for this purpose. You can opt-in to share data.
>2. Any data sent through the API will be retained for abuse and misuse monitoring purposes for a maximum of 30 days, after which it will be deleted (unless otherwise required by law).

**Comment by jlyons210** [comment]
I'm going to keep this issue open for feedback on this feature, as 0.7.0 is pretty experimental and I'm sure tuning and refinement will be necessary.

**Comment by jlyons210** [comment]
Issues I've noticed so far:
- 20% is way too high of a default for auto-react/engage, 5% has been working out better,
- The bot's _own_ messages are excluded from auto-engagement, but _other_ bots' messages are not. Bots can create a conversation storm,
- 2 is too many bot reactions at once.

I implemented the conversation auto-engagement pretty lazily, partly out of frustration with how I am holding ephemeral data. I noticed other gripes with the code organization that I'd like to make better:
- Passing big objects to functions - rather than exactly what is needed,
- Too much Discord-specific logic is in `app.js` rather than `lib-discord.js` and needs to be passed back and forth (should be contained within the module,
- The theme is that there is a lot to gain by properly implementing proper OOP.

Reducing the complexity of handling data will make this a lot easier to add features to.

**Comment by jlyons210** [comment]
Issues I've noticed so far:
- 20% is way too high of a default for auto-react/engage, 5% has been working out better,
- The bot's _own_ messages are excluded from auto-engagement, but _other_ bots' messages are not. Bots can create a conversation storm,
- 2 is too many bot reactions at once,
- Bot reactions should look at conversation context rather than just single messages to react to.

I implemented the conversation auto-engagement pretty lazily, partly out of frustration with how I am holding ephemeral data. I noticed other gripes with the code organization that I'd like to make better:
- Passing big objects to functions - rather than exactly what is needed,
- Too much Discord-specific logic is in `app.js` rather than `lib-discord.js` and needs to be passed back and forth (should be contained within the module,
- The theme is that there is a lot to gain by properly implementing proper OOP.

Reducing the complexity of handling data will make this a lot easier to add features to.
